Cleanup some leaked file handles

weirddan455 opened the pull request:
# Description

This allows Dunkle Schatten 2 to load and past the first couple of screens before it freezes.  For some reason, it's very sensitive to extra files being created.  It's kind of like a DOS Valgrind because it helped me clean up some leaks.

Once it gets past a certain number of files, it stomps on our device chain and then spins forever in this loop:

https://github.com/dosbox-staging/dosbox-staging/blob/ffea16b2a60dbcb545654246e1d67d7af9eb55aa/src/dos/dos_devices.cpp#L238-L257

I haven't fixed the root problem yet but it lead me to some buggy code I could fix regardless.

## Related issues

#3668


# Manual testing

- Output redirection works `dir > test.txt`
- Pipe works `dir | more`
- Input redirection works `test.txt < more`

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -131,47 +133,41 @@ void DOS_Shell::GetRedirection(char *line,
 }
 
-bool get_pipe_status(const char *out_file,
+static uint16_t get_output_redirction(const char *out_file,
```
Typo: redirection

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -256,23 +214,18 @@ void DOS_Shell::ParseLine(char *line)
+	const uint8_t old_stin = psp->GetFileHandle(STDIN);
+	const uint8_t old_stdout = psp->GetFileHandle(STDOUT);
+
+	uint16_t input_redirction = InvalidFileHandle;
```
Typo: redirection

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -256,23 +214,18 @@ void DOS_Shell::ParseLine(char *line)
+	const uint8_t old_stdout = psp->GetFileHandle(STDOUT);
+
+	uint16_t input_redirction = InvalidFileHandle;
+	uint16_t output_redirection = InvalidFileHandle;
```
Use auto?

weirddan455 replied:
auto'd

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -256,23 +214,18 @@ void DOS_Shell::ParseLine(char *line)
-	bool normalstdin = false;  /* whether stdin/out are open on start. */
-	bool normalstdout = false; /* Bug: Assumed is they are "con"      */
+	const uint8_t old_stin = psp->GetFileHandle(STDIN);
+	const uint8_t old_stdout = psp->GetFileHandle(STDOUT);
```
Use auto?

weirddan455 replied:
auto'd

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -308,49 +261,44 @@ void DOS_Shell::ParseLine(char *line)
-			open_console_device();
+		output_redirection = get_output_redirction(out_file.length() ? out_file.c_str() : nullptr,
+		                                           pipe_file.length() ? pipe_file.c_str() : nullptr,
+												   pipe_tempfile, append, failed_pipe);
```
Indentation / formatting

weirddan455 replied:
Formatted

weirddan455 commented:
I've addressed your comments and tested redirection.  I had a bug with pipes that I fixed.

PVS was throwing some false positive warnings so I did some `#pragma pvs` to disable it but now MSVC is throwing warnings about unknown pragmas.  Stupid build systems man... don't worry about it MSVC that pragma is not meant for you.

weirddan455 commented:
CI all passes now finally.  I think this is good to go but I'll give you a chance to look it over again if you want @johnnovak since I made some changes.

johnnovak commented on src/shell/file_reader.cpp:
```diff
@@ -29,7 +29,7 @@ std::optional<FileReader> FileReader::GetFileReader(const std::string& filename)
 	}
 	DOS_CloseFile(handle);
-	return FileReader(std::move(fullname));
+	return std::unique_ptr<FileReader>(new FileReader(std::move(fullname)));
```
Isn't `return new FileReader...` enough here?

weirddan455 replied:
Doesn't compile.  `std::make_unique` doesn't work either because it's a private constructor so I think this is the best way.

```
../../src/shell/file_reader.cpp:32:9: error: no viable conversion from returned value of type 'FileReader *' to function return type 'std::unique_ptr<FileReader>'
        return new FileReader(std::move(fullname));
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

johnnovak replied:
Sure.

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -131,47 +133,41 @@ void DOS_Shell::GetRedirection(char *line,
+			success = DOS_CreateFile(out_file,
 			                        FatAttributeFlags::Archive,
-			                        &dummy);
+			                        &file_handle, true);
```
What is `true`? Extract to constant please.

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -131,47 +133,41 @@ void DOS_Shell::GetRedirection(char *line,
-		                              &dummy2);
-		if (pipe_file && (failed_pipe || !status) &&
+		success = DOS_CreateFile(pipe_file && !failed_pipe ? pipe_tempfile : out_file,
+		                         FatAttributeFlags::Archive, &file_handle, true);
```
What is `true`? Extract to constant please.

johnnovak replied:
Maybe put parens around `(pipe_file && !failed_pipe)`, or extract it to a temp var.

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -187,55 +183,18 @@ bool get_pipe_status(const char *out_file,
-				                              &dummy,
-				                              &dummy2);
+				if (success) {
+					DOS_CloseFile(file_handle, true);
```
What is `true`?

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -187,55 +183,18 @@ bool get_pipe_status(const char *out_file,
+				if (success) {
+					DOS_CloseFile(file_handle, true);
+				}
+				success = DOS_CreateFile(pipe_tempfile, FatAttributeFlags::Archive, &file_handle, true);
```
`true`... 😄

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -256,23 +215,18 @@ void DOS_Shell::ParseLine(char *line)
-		if ((dummy = open_stdin_as(in_file.c_str())) != failed_open) { // Test if
-			// file exists
-			DOS_CloseFile(dummy);
+		if (DOS_OpenFile(in_file.c_str(), OPEN_READ, &input_redirection, true)) {
```
true

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -308,49 +262,57 @@ void DOS_Shell::ParseLine(char *line)
 				                           : "(unnamed)");
-				close_stdout();
-				open_stdout_as("nul");
+				DOS_OpenFile("nul", OPEN_READWRITE, &output_redirection, true);
```
true

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -308,49 +262,57 @@ void DOS_Shell::ParseLine(char *line)
-		open_console_device(normalstdin);
+	if (input_redirection != InvalidFileHandle) {
+		psp->SetFileHandle(STDIN, old_stdin);
+		DOS_CloseFile(input_redirection, true);
```
I don't even need to say anything 😄

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -308,49 +262,57 @@ void DOS_Shell::ParseLine(char *line)
-		close_stdin(!normalstdin);
+	if (output_redirection != InvalidFileHandle) {
+		psp->SetFileHandle(STDOUT, old_stdout);
+		DOS_CloseFile(output_redirection, true);
```
...

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -308,49 +262,57 @@ void DOS_Shell::ParseLine(char *line)
+			safe_strcpy(mutable_pipe_file, pipe_file.c_str());
+			ParseLine(mutable_pipe_file);
+			psp->SetFileHandle(STDIN, old_stdin);
+			DOS_CloseFile(pipe_handle, true);
```
`true` 😉 and please insert some newlines while you're at it, this block of ASCII is absolutely unreadable.

weirddan455 replied:
Fixed

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -202,6 +202,11 @@ uint16_t get_tick_random_number() {
+// Disable PVS warning 1020 for this function.
+// It's a false positive about needing to call DOS_CloseFile()
+#pragma pvs(push)
+#pragma pvs(disable: 1020)
```
I'd suggest disabling the warning only for that single line with putting a `//-V1020` comment at the end of it.

Just run `git grep '//-V'` on our repo, we're using that form of suppression in a few places.

weirddan455 replied:
That's the way I usually like to do it but it's throwing 4 separate instances of that error on different lines and on lines that only contain a closing bracket.

![pvs](https://github.com/dosbox-staging/dosbox-staging/assets/8128047/3ec4637e-e324-48d7-8a49-16d84a04215b)

johnnovak replied:
Yeah ok, if you can't refactor it somehow so the warning appears on a single line only, then the pragma approach is the better way, indeed.

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -202,6 +202,11 @@ uint16_t get_tick_random_number() {
 }
 
+// Yo dawg MSVC-Clang likes to complain about pragmas
+// So here's a pragma so you stop complaining about pragmas
```
Funny 😄 But if you do the per-line suppression thing by the magic PVS-Studio specific comment, you won't need this 😄

johnnovak reviewed (approved):
Still looking good 😄 A few minor comments, but good job otherwise @weirddan455 👍🏻 

I found this amusing, btw:

>  It's kind of like a DOS Valgrind because it helped me clean up some leaks.

johnnovak commented on src/shell/shell.cpp:
```diff
@@ -251,28 +221,25 @@ void DOS_Shell::ParseLine(char *line)
 		line[0] = ' ';
 	line = trim(line);
 
+	constexpr bool fcb = true;
```
You've done enough @weirddan455, so you're off the hook... But WTF is `fcb`? Job for the next guy who touches this file to figure out, i guess 😆

weirddan455 replied:
https://en.wikipedia.org/wiki/File_Control_Block
